pkg/datastore/headerscontroller.go

```diff
@@ -0,0 +1,282 @@
+	if l > 1 {
+		o := hc.ordered[l-1]
+		if o.Slot >= i.Slot {
+			sort.Slice(hc.ordered, func(i, j int) bool {
+				return hc.ordered[i].Slot < hc.ordered[j].Slot
+			})
+		}
+	}
+}
+
+func (hc *HeaderController) GetLatestSlot() (slot uint64) {
+	return atomic.LoadUint64(hc.latestSlot)
+}
+
+func (hc *HeaderController) AddMultiple(slot uint64, hnt []structs.HeaderAndTrace) (err error) {
+	hc.cl.Lock()
+	defer hc.cl.Unlock()
+
+	h, ok := hc.content[slot]
+	if !ok {
+		h = NewHNTs()
+		hc.content[slot] = h
+		hc.addToOrdered(Info{
+			Slot:  slot,
+			Added: time.Now(),
+		})
+
+		ls := atomic.LoadUint64(hc.latestSlot)
+		if ls == 0 || slot > ls {
+			atomic.StoreUint64(hc.latestSlot, slot)
+		}
+	}
+
+	for _, s := range hnt {
+		if err := h.AddContent(s); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (hc *HeaderController) Add(slot uint64, hnt structs.HeaderAndTrace) (newCreated bool, err error) {
+	hc.cl.Lock()
+	defer hc.cl.Unlock()
+
+	h, ok := hc.content[slot]
+	if !ok {
+		h = NewHNTs()
+		hc.content[slot] = h
+		hc.addToOrdered(Info{
+			Slot:  slot,
+			Added: time.Now(),
+		})
+		newCreated = true
+
+		ls := atomic.LoadUint64(hc.latestSlot)
+		if ls == 0 || slot > ls {
+			atomic.StoreUint64(hc.latestSlot, slot)
+		}
+	}
+	return newCreated, h.AddContent(hnt)
+}
+
+func (hc *HeaderController) GetContent(startingSlot, stopSlot uint64, limit int) (elements []structs.HeaderAndTrace, lastSlot uint64) {
+	for _, o := range hc.GetOrderedDesc() {
+		if o > startingSlot || o < stopSlot {
+			continue
+		}
+		hc.cl.RLock()
+		h, ok := hc.content[o]
+		hc.cl.RUnlock()
+		if !ok || h == nil {
+			continue
+		}
+		lastSlot = o
+		c, _ := h.GetContent()
+		elements = append(elements, c...)
+		if len(elements) >= limit {
+			return elements, lastSlot
+		}
+	}
+
+	return elements, lastSlot
+}
+
+func (hc *HeaderController) GetSingleSlot(slot uint64) (elements []structs.HeaderAndTrace, revision uint64, err error) {
+	hc.cl.RLock()
+	h, ok := hc.content[slot]
+	hc.cl.RUnlock()
+
+	if !ok {
+		return nil, 0, nil
+	}
+	elements, revision = h.GetContent()
+	return elements, revision, err
+}
+
+func (hc *HeaderController) UnlinkElement(slot, expectedRevision uint64) (success bool) {
+	hc.cl.Lock()
+	defer hc.cl.Unlock()
+
+	s, ok := hc.content[slot]
+	if !ok {
+		return true
+	}
+
+	if expectedRevision != s.GetRevision() {
+		return false
+	}
+
+	// we're only unlinking from map here
+	// it should be later on eligible for GC
+	delete(hc.content, slot)
+
+	// removed element should be first on the left
+	if len(hc.ordered) > 0 {
+		for i, v := range hc.ordered {
+			if v.Slot == slot {
+				hc.ordered = append(hc.ordered[:i], hc.ordered[i+1:]...)
+				break
+			}
+		}
+	}
+
+	return true
+}
+
+func (hc *HeaderController) GetMaxProfit(slot uint64) (hnt structs.HeaderAndTrace, ok bool) {
+	hc.cl.RLock()
+	defer hc.cl.RUnlock()
+
+	s, ok := hc.content[slot]
+	if !ok {
+		return hnt, false
+	}
+	return s.GetMaxProfit()
+
+}
+
+type HNTs struct {
```

Comment:
What does HNTs mean? We should use more descriptive name for this

Reply:
"Header And Trace" -> "Header'N'Trace" -> "HNT" . There are a lot wrappers of the basic structure. 
I didn't have a better idea. `SlotContents` ?

Reply:
What about `OptimizedHeaders` or `IndexedHeaders`?

Also why do you use `uint64` instead of `Slot`? Any reason?

Reply:
As much as I see the only place where it has anything called from it is `Epoch()` method  inside Service.
It's never really used in any operation, and it's simply just a number. Especially as we grow more into modules, it would be significantly easier to operate on base types, as this doesn't carry any additional value with it. 

`IndexedHeaders` will do.

Reply:
renamed